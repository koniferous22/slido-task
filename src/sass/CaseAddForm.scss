@use 'base';

form#case-add {
	@include base.page-section(auto, 600px);
	display: grid;
	grid-template-columns: base.$form-label-column-width auto;
	/*margin: auto;
	max-width: 600px;*/
	width: base.$form-width;

	%form-full-width {
		grid-column: 1 / (base.$form-columns + 1);
		text-align: center;
	}

	h3#form-title {
		@extend %form-full-width;
	}
	
	div.error {
		@extend %form-full-width;
		font-weight: bold;
		color: red;
	}

	input, textarea {
		border-radius: base.$form-input-border-radius;
	}

	textarea {
		resize: none;
	}
	
	input.submit-button {
		@extend %form-full-width;
		margin: 20px auto;
		border: base.$form-submit-border;
		
		background-color: transparent;
		font-weight: bold;
	}

	input.submit-button:hover {
		text-decoration: underline;
		cursor: pointer;
	}
}
