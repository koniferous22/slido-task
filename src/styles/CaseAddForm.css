form#case-add {
	grid-area: form;

	display: grid;
	grid-template-columns: 100px auto;
	grid-gap: 5px 5px;

	margin: auto;
	width: 66vw;
	max-width: 600px;
}

form#case-add label {
	justify-self: end;
}

form#case-add h3#form-title {
	grid-column: 1/3;
	text-align: center;
}

form#case-add div.error {
	grid-column: 1/3;
	text-align: center;
	font-weight: bold;
	color: red;
}

form#case-add input.submit-button {
	grid-column: 1/3;
	text-align: center;

	background-color: transparent;
	
	margin: 20px auto 10px auto;
	border-radius: 3px;
	border: 2px solid #51300a;

	font-weight: bold;
}

form#case-add input.submit-button:hover {
	text-decoration: underline;
	cursor: pointer;
}
