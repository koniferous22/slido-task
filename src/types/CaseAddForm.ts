import { Case } from 'types/common/Case'
import { FormRecord, FormMaybeRecord } from 'types/form/FormRecord'

export interface CaseAddFormState {
	errorFlags: FormMaybeRecord<boolean>;
	values: FormRecord<string>;
}

export interface CaseAddFormProps {
	onSubmit: (case_: Case) => void;
}