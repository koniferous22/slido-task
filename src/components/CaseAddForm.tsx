import React from 'react'

import { Case } from 'types/common/Case'
import { FormRecord, FormMaybeRecord } from 'types/form/FormRecord'
import { CaseAddFormProps, CaseAddFormState } from 'types/CaseAddForm'

import FormInput from './form/FormInput'
import FormTextArea from './form/FormTextArea'

import 'styles/CaseAddForm.css'

const DEFAULT_FORM_VALUES: FormRecord<string> = {
	title: '',
	description: '',
	date: ''
}

class CaseAddForm extends React.Component<CaseAddFormProps, CaseAddFormState> {
	constructor(props: CaseAddFormProps) {
		super(props)

		this.state = {
			values: { ...DEFAULT_FORM_VALUES },
			errorFlags: {}
		}
		this.handleFormChangeEvent = this.handleFormChangeEvent.bind(this)
		this.handleFormBlurEvent = this.handleFormBlurEvent.bind(this)
		this.handleSubmit = this.handleSubmit.bind(this)
	}

	validateFormFields(...fields: Array<keyof Case>): void {
		const reduceCb = (errorFlags: FormMaybeRecord<boolean>, field: keyof Case): FormMaybeRecord<boolean> => {
			return Object.assign(errorFlags, {[field]: (this.state.values[field].trim() === '')})
		}
		const newFormErrorFlags = fields.reduce(reduceCb, this.state.errorFlags)
		this.setState({errorFlags: newFormErrorFlags})
	}

	handleFormBlurEvent(field: keyof Case): void {
		this.validateFormFields(field)
	}

	handleFormChangeEvent(field: keyof Case, eventValue: string): void {
		this.setState({
			values: Object.assign(this.state.values, {[field]: eventValue})
		})
	}

	handleSubmit(event: React.FormEvent<HTMLFormElement>): void {
		event.preventDefault();
		
		this.validateFormFields('title', 'description', 'date')

		const { values, errorFlags } = this.state

		// verify that every error has been fixed
		const formKeys: Array<keyof Case> = Object.keys(errorFlags) as Array<keyof Case>
		const noFormError: boolean = formKeys.every((key: keyof Case) => errorFlags[key] === false )
		if (noFormError) {
			const newCase: Case = {
				title: values.title || '',
				description: values.description || '',
				date: values.date ? new Date(values.date) : new Date()
			}
			this.setState({ values: { ...DEFAULT_FORM_VALUES }})
			this.props.onSubmit(newCase)
		}
	}

	render() {

		const { errorFlags, values } = this.state

		return (
			<form id="case-add" onSubmit={this.handleSubmit}>
				<h3 id="form-title">Submit event here</h3>
				<FormInput
					name="title" label="Title"
					type="text"
					errorFlag={Boolean(errorFlags.title)}
					onBlur={this.handleFormBlurEvent}
					onChange={this.handleFormChangeEvent}
					value={values.title || ''}
				/>
				<FormTextArea
					name="description" label="Description"
					errorFlag={Boolean(errorFlags.description)}
					onBlur={this.handleFormBlurEvent}
					onChange={this.handleFormChangeEvent}
					rows={5}
					value={values.description || ''}
				/>
				<FormInput
					name="date" label="Event Date"
					type="date"
					errorFlag={Boolean(errorFlags.date)}
					onBlur={this.handleFormBlurEvent}
					onChange={this.handleFormChangeEvent}
					value={values.date || ''}
				/>
				<input className="submit-button" type="submit" value="Add Event" />
			</form>
		)
	}
}


export default CaseAddForm