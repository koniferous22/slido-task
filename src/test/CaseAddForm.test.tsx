import React from 'react'
import renderer from 'react-test-renderer'
import CaseAddForm from 'components/CaseAddForm'

it('CaseAddForm renders correctly', () => {
	const tree = renderer
		.create(<CaseAddForm onSubmit={(): void => {return;}}/>)
		.toJSON();
	expect(tree).toMatchSnapshot();
});
