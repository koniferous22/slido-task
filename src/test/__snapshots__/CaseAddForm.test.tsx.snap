// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CaseAddForm renders correctly 1`] = `
<form
  id="case-add"
  onSubmit={[Function]}
>
  <h3
    id="form-title"
  >
    Submit event here
  </h3>
  <label
    htmlFor="title"
  >
    Title: 
  </label>
  <input
    id="title"
    name="title"
    onBlur={[Function]}
    onChange={[Function]}
    type="text"
    value=""
  />
  <label
    htmlFor="description"
  >
    Description: 
  </label>
  <textarea
    id="description"
    name="description"
    onBlur={[Function]}
    onChange={[Function]}
    rows={5}
    value=""
  />
  <label
    htmlFor="date"
  >
    Event Date: 
  </label>
  <input
    id="date"
    name="date"
    onBlur={[Function]}
    onChange={[Function]}
    type="date"
    value=""
  />
  <input
    className="submit-button"
    type="submit"
    value="Add Event"
  />
</form>
`;
